src/include/misc.h

```diff
@@ -309,7 +309,7 @@
 static inline bool
 __wt_string_match(const char *str, const char *bytes, size_t len)
 {
-    return (len > 0 && (strlen(str) == len) && (strncmp(str, bytes, len) == 0));
+    return (strncmp(str, bytes, len) == 0 && str[len] == '\0');
```

Comment:
I do not believe `len==0` is handled correctly in this version.
First if `len==0` then `strncmp` will evaluate as true.

Then there are 3 possible outcomes:
1. `str` points to a `\0` character and match == true. VALID
2. `str` points to a non `\0` character and match == false. INVALID
3. `str` points "somewhere" so undefined behaviour. INVALID and may crash.

A possible fix is to short circuit the `len==0` case:
```suggestion
    return (len == 0 || (strncmp(str, bytes, len) == 0 && str[len] == '\0'));
```
I
 understand this is consistent as `strncmp` with a `len==0` will return 0. 
Where `len < strlen(str)` accessing `str[len]` should be valid.
Where `len > strlen(str)` the `strncmp` should fail and short circuit.

While I would be very surprised if adding `len==0` impacted performance: but I still suggest it is appropriate to benchmarking any/all solutions.

If you'd like to try @sueloverso suggestion you could slip it in before `strncmp`:
```suggestion
    return (len == 0 || (*str == *bytes && strncmp(str, bytes, len) == 0 && str[len] == '\0'));
```

I think we could benefit from the `restrict` keyword here -- but afaik it is not support by MSVC.

Reply:
I think @marcbutler is right.  Once upon a time, this macro was used in limited cases where we knew certain things about the args (str and bytes were always valid memory even if length was zero, first arg was typically a literal string, e.g. "foo", we never had a situation where both sides were len 0, etc.). Now this macro is used more generally, so the `len == 0` makes sense as protection.

The main issue I have with the committed version is the strlen call - the first string would be scanned twice in the general case.

I am not strong either way on the initial `*str == *bytes` check.  It was in the original version of this macro, but I think compilers and processors are getting smarter all the time and may know to do this (and possibly know better ways, or tuned to the specific architecture).  The only way to know is to measure.

Reply:
Thanks. It makes sense to me. I did some perf testing and `*str == *bytes` is not making any meaningful difference on the perf results. I'll leave it aside for now.